### PR TITLE
Restructure domain classes

### PR DESCRIPTION
**Pull request by lorenzleutgeb**
New interfaces:
- `interface Identifiable<T>` that is a `@FunctionalInterface` of `T getId()` to be used for anything (Domain classes and DTOs that have an ID).
- `interface Canonical<T> extends Identifiable<T>` that adds `String getCanonicalName()` to be used for anything (Domain classes and DTOs that have an ID, and a canonical name).
- `interface Named<T> extends Canonical<T>` that adds `String getName()` to be used for anything (Domain classes and DTOs that have an ID, a canonical name, and a name).

New abstract implementations:
- `abstract class IdentifiableEntity implements Identifiable<UUID>`
- `abstract class CanonicalEntity extends IdentifiableEntity implements Canonical<UUID>`
- `abstract class NamedEntity extends CanonicalEntity implements Named<UUID>`

New hierarchy:
- `Assignment` , `Challenge`, `Organization` and `Team` extend `NamedEntity`.
- `Language` now also `extends NamedEntity` and uses a canonicalname instead of a "tag".
- `User extends CanonicalEntity` (username is the same as canonical name) that also implements `Named<UUID>` (see `User.getName()`).

Inconsistencies:
- `Endpoint` has a name and `getName()` but does not implement `Named` because it has no canonical name, this might be misleading.

Also, I added constructors to domain classes in many places. When this get's merged, the DTO hierarchy will be improved as well.

### CODE REVIEW

File: platform-server-kernel/src/main/resources/i18n/exceptions.properties [hunk 1]
```diff
@@ -52,10 +52,10 @@ ccc.game.structure.unsuported=Importing games with external validation classes i
...
-location.name.existing=Location with name `{0}` already exists
+location.canonicalName.existing=Location with canonicalName `{0}` already exists
 location.id.existing=There is already a location created for id `{0}` named `{1}`
 
 participation.invalid = The participation you provided is invalid
 participation.invitation.invalid = The participation invitation is invalid
 
-challenge.template.canonical.name.exists=This canonical name already exists for a challenge template.
\ No newline at end of file
+challenge.template.canonical.canonicalName.exists=This canonical canonicalName already exists for a challenge template.
```
Comment (Exordian) [inline platform-server-kernel/src/main/resources/i18n/exceptions.properties]: `canonical canonicalName` absicht?

File: platform-server-kernel/src/main/resources/i18n/exceptions_en_US.properties [hunk 2]
```diff
@@ -4,7 +4,7 @@ task.denied=Access denied to task
 challenge.invalid=Invalid challenge
 challenge.ended=Challenge has ended
 challenge.completed=Challenge already completed
-challenge.canonicalName.existing=A challenge with `{0}` canonical name already exisits
+challenge.canonicalName.existing=A challenge with `{0}` canonical canonicalName already exisits
```
Comment (Exordian) [inline platform-server-kernel/src/main/resources/i18n/exceptions_en_US.properties]: same

File: platform-server-kernel/src/main/resources/i18n/exceptions_en_US.properties [hunk 3]
```diff
@@ -32,8 +32,8 @@ email.existing=Email `{0}` already exists
 password.old.invalid=Old password does not match
 password.new.match=New password does identical to old one
 
-team.canonicalName.existing = Team tag name already exists
-team.name.existing=Team name already exists
+team.canonicalName.existing = Team name canonicalName already exists
```
Comment (Exordian) [inline platform-server-kernel/src/main/resources/i18n/exceptions_en_US.properties]: same

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/domain/Language.java [hunk 4]
```diff
@@ -5,38 +5,30 @@
 import javax.persistence.Table;
 import java.util.Collections;
 import java.util.Set;
+import java.util.UUID;
 
 @Entity
 @Table(name = "language")
-public class Language extends IdentifiableEntity {
-    private String name;
-    private String tag;
```
Comment (Exordian) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/domain/Language.java:13]: migration for `tag` missing?

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/language/LanguageShowDto.java [hunk 5]
```diff
@@ -7,8 +7,8 @@
 
 public class LanguageShowDto {
     private UUID id;
+    private String canonicalName;
```
Comment (Exordian) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/language/LanguageShowDto.java:10]: frontend change missing

Comment (lorenzleutgeb) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/language/LanguageShowDto.java:10]: Frontend change is at coduno/app#333

**Pull request merged**